### ISSUE TITLE
Image upload leaves imageSizes blank

### ISSUE THREAD

**Issue by HarleySalas** [payloadcms/payload#3915]
### Image upload leaves imageSizes blank:

_Note: the image upload I'm referring to here specifically is 1024x1024 resolution. (It also happens with all other sizes though, when there is a large disparity between source image width and resize `imageSizes` dimensions._

![image](https://github.com/payloadcms/payload/assets/38236585/1593c8b8-e65c-45b8-8894-366c86505aa7)

**Media Upload in Collection:**
```
upload: {
    staticURL: "/media",
    staticDir: "media",
    crop: true,
    focalPoint: true,
    formatOptions: {
      format: "webp",
      options: {
        quality: 85,
      },
    },
    imageSizes: [
      {
        name: "blur",
        width: 32,
        //revert to 32 when #3859 is released (force image generation when aspect ratio matches)
        height: 30,
        formatOptions: {
          format: "png",
        },
      },
      {
        name: "thumbnail",
        width: 400,
        height: 300,
        formatOptions: {
          format: "webp",
        },
      },
      {
        name: "preview",
        width: 10,
        height: 10,
        formatOptions: {
          format: "webp",
        },
      },
      {
        name: "landscape",
        width: 1440,
        height: 1080,
        formatOptions: {
          format: "webp",
        },
      },
      {
        name: "portrait",
        width: 1080,
        height: 1440,
        formatOptions: {
          format: "webp",
        },
      },
    ],
    adminThumbnail: "thumbnail",
    mimeTypes: ["image/*"],
  },
  ```
  
I don't believe this should be possible. I'm not sure if it's a bug, or if this behavior is intentional.

Let's say I have an image that is 1920x1080.
I have an `ImageSize` like this:
```
{
	name: "square",
	width: 3000,
	height: 3000,
}
```
Currently, it will not put any image for `"square"`, as the width property is far larger than the width on my original image.

In this instance, I believe it should take note of the aspect ratio in my `ImageSize`, in this case **1:1**. It should then take the largest dimension of my source image and create the resized image with that; the outputted image under `"square"` should be 1920x1920, in my opinion.

There is no scenario where upscaling an image will be beneficial for anyone. I guess it makes sense to have an *option* for it, but it is not a good best practice, at all. What's just as bad, or worse is getting undefined values from your api. 

### If that solution, or behavior is not preferred, I think a real world scenario like this should be considered:

I'm using Next Image. I prefer for it to create and handle all of my resizing/srcset creation. However, I want the benefits of focal points and aspect ratios from an image upload feature.

Sometimes, the height and width of an image are unimportant. Not always, but sometimes they are. In my instance, all I care about is displaying something in a 1:1, or 3:4 aspect ratio, but I have no idea what aspect ratio my client will upload.

An acceptable solution would be to add a field to `ImageSize` called `aspectRatio` where I could enter "1:1" or "16:9", "9:16", etc and it would simply make the largest possible version of the uploaded image, while respecting the original image's height/width and focal point indicated. This way, I can simply access the image I need with `image.sizes.portrait.url`, or `image.sizes.landscape.url`, for example. The original width/height properties could be no longer required, they could be used as the maximum dimensions allowed for the resize, or the most explicit way to do it would be to not allow width/height properties at all, if `aspectRatio` is provided and then add additional properties called `maxWidth` and `maxHeight`, so that the operation taking place is impossible to misunderstand.

### To Reproduce

use this in an upload collection:

```
upload: {
    staticURL: "/media",
    staticDir: "media",
    crop: true,
    focalPoint: true,
    formatOptions: {
      format: "webp",
      options: {
        quality: 85,
      },
    },
    imageSizes: [
      {
        name: "blur",
        width: 32,
        height: 30,
        formatOptions: {
          format: "png",
        },
      },
      {
        name: "thumbnail",
        width: 400,
        height: 300,
        formatOptions: {
          format: "webp",
        },
      },
      {
        name: "preview",
        width: 10,
        height: 10,
        formatOptions: {
          format: "webp",
        },
      },
      {
        name: "landscape",
        width: 1440,
        height: 1080,
        formatOptions: {
          format: "webp",
        },
      },
      {
        name: "portrait",
        width: 1080,
        height: 1440,
        formatOptions: {
          format: "webp",
        },
      },
    ],
    adminThumbnail: "thumbnail",
    mimeTypes: ["image/*"],
  },
  ```
  
  upload the following image:
  
![about-help-1-v2](https://github.com/payloadcms/payload/assets/38236585/7894e4ab-9e6c-442f-a57b-d7b4444a2a08)

go to **Preview Sizes** and note that no landscape, or portrait options are generated.

### Payload Version

2.0.13

**Comment by JessChowdhury** [comment]
Hi @HarleySalas - thank you for the clear description here! 

> There is no scenario where upscaling an image will be beneficial for anyone. I guess it makes sense to have an option for it, but it is not a good best practice, at all.

Due to the way [Sharp](https://sharp.pixelplumbing.com/api-resize) works, the default behavior here should be that the image gets enlarged or reduced based on the dimensions provided (even though it seems not ideal). You would then pass a prop for `withoutEnlargement` or `withoutReduction` if you don't want the uploaded image to be enlarged / reduced.

> I believe it should take note of the aspect ratio in my `ImageSize`

We would likely want to add a fallback option for this rather than doing it by default. I definitely agree that it would be useful to have an aspect ratio option - as you said, there are scenarios where the width/height is not important and being able to define an aspect ratio would provide a solution here.

So, there are 2 parts to this...

1. Fixing the current issues - blank image sizes in the preview drawer and returning null image values
2. Adding the aspect ratio option - which would be a new feature 

I will aim to resolve the issues ASAP and we can get the aspect ratio feature on our roadmap for the near future - how does that sound?

**Comment by HarleySalas** [comment]
Sounds good to me. I appreciate the quick effort 👍 

To clarify though, once its fixed;
If I set `withoutEnlargement`, it will basically work as I originally thought it would? In the sense that if I have a 3000x3000 resize target, `withoutEnlargement` is `true` and I upload a 1500x2000 image, it will still provide me with a 1500x1500 image, for example?

If so, that is perfectly fine enough for me and I'd say the aspect ratio option would just be a quite low priority nice-to-have.

**Comment by JessChowdhury** [comment]
> If I set withoutEnlargement, it will basically work as I originally thought it would? In the sense that if I have a 3000x3000 resize target, withoutEnlargement is true and I upload a 1500x2000 image, it will still provide me with a 1500x1500 image, for example?

Yes! This is correct. 

Also, I previously said that _'the default behavior here should be that the image gets enlarged or reduced_' but then found that Payload is manually preventing the default enlargement. This is the reason you are received the `null` images sizes, instead of seeing enlarged ones by default. Currently `withoutEnlargement: false` is needed in your `imageSize` for enlargement to happen - I know enlargement isn't your desired behavior but just wanted to clear it up.

**Comment by HarleySalas** [comment]
So far, I haven't gotten any blank images, which is great since the release. However, withoutEnlargement doesn't appear to be working properly? I've tried to make sense of your previous comments and I think I've understood correctly, but at the same time, I'm not getting the desired outcome, so perhaps I didn't understand 😅

I have the following:

I uploaded a 960 x 1280 .webp image, with these `imageSizes`:
```
    {
        name: "thumbnail",
        width: 400,
        height: 300,
      },
      {
        name: "preview",
        width: 800,
        height: 800,
        formatOptions: {
          format: "webp",
        },
        withoutEnlargement: true,
      },
      {
        name: "landscape",
        width: 2880,
        height: 2160,
        formatOptions: {
          format: "webp",
        },
        withoutEnlargement: true,
      },
      {
        name: "portrait",
        width: 2160,
        height: 2880,
        formatOptions: {
          format: "webp",
        },
        withoutEnlargement: true,
      },
      {
        name: "square",
        width: 2160,
        height: 2160,
        formatOptions: {
          format: "webp",
        },
        withoutEnlargement: true,
      },
```
Result:
thumbnail - 400x300: **good**
preview - 800x800 **good**
landscape - 2880x2160 **bad**: should be maximum 960 wide
portrait - 960x1280 **good** (note: same size as source image)
square - 2160x2160 **bad**

If I switch `withoutEnlargement` to false, `portrait` then becomes 2160x2880, so I guess that makes sense. We are telling it to proceed with enlargement. In the case of setting it to true, as you can see, it is still enlarging the images.

I *think* I've understood that you may have commented about this behavior, but you were also mentioning the aspect ratio feature in the future as well? It's not 100% clear to me. I'd just like to make sure we're on the same page, that you're aware of current behavior and that *hopefully* at least `withoutEnlargement` set to true will work as expected; no enlargement **and** no null image. It seems to me that as of now, it's just forcing enlargement to guarantee that **specific** dimension is provided. But that is strange! why wouldn't it **also** enlage `portrait` as well under those conditions?

**Comment by JarrodMFlesch** [comment]
Hey @HarleySalas I have some fixes for ya in #4039!

Thank you for hanging in there and providing feedback 😄
